src/storage/page_buffer.c

```diff
@@ -8489,22 +8517,121 @@ pgbuf_respond_data_fetch_page_request (THREAD_ENTRY &thread_r, std::string &payl
       // pack NO_ERROR first
       payload_in_out = { reinterpret_cast<const char *> (&error), sizeof (error) };
 
+      const char *compressed_data = nullptr;
+      int compressed_data_size = 0;
+      const bool compressed = pgbuf_compress_page(thread_r, reinterpret_cast<const void *> (io_pgptr), IO_PAGESIZE,
+                                                  compressed_data, compressed_data_size);
+      assert (compressed);
+
+      if (!compressed)
+        {
+          // TODO: if compression failed, set error and send original page
+          assert_release (false);
+        }
+
+      // add compressed size
+      const PGLENGTH compressed_length = (PGLENGTH)compressed_data_size;
+      payload_in_out.append (reinterpret_cast<const char *> (&compressed_length),
+                             sizeof (PGLENGTH));
+
       // add io_page
-      payload_in_out.append (reinterpret_cast<const char *> (io_pgptr), (size_t) IO_PAGESIZE);
+      payload_in_out.append (compressed_data, (size_t) compressed_data_size);
 
       if (prm_get_bool_value (PRM_ID_ER_LOG_READ_DATA_PAGE))
 	{
 	  _er_log_debug (ARG_FILE_LINE,
 			 "[READ DATA] Successful while fixing page: vpid = %d|%d, lsa = %lld|%d,"
-			 " target_repl_lsa = %lld|%d\n",
-			 VPID_AS_ARGS (&vpid), LSA_AS_ARGS(&io_pgptr->prv.lsa), LSA_AS_ARGS (&target_repl_lsa));
+			 " target_repl_lsa = %lld|%d\n"
+	                 "    compr_size = %d, ratio = %.2lf\n",
+			 VPID_AS_ARGS (&vpid), LSA_AS_ARGS(&io_pgptr->prv.lsa), LSA_AS_ARGS (&target_repl_lsa),
+	                 (int) compressed_data_size, ((double)compressed_data_size/(double)IO_PAGESIZE) * 100.);
 	}
 
       // only unfix after having used casted IO page ptr for logging
       pgbuf_unfix (&thread_r, page_ptr);
     }
 }
 // *INDENT-ON*
+
+bool
+pgbuf_compress_page (THREAD_ENTRY & thread_r, const void *data_in, int data_in_size,
+		     const char *&data_out, int &data_out_size)
+{
+  assert (data_in != nullptr && data_in_size > 0);
+
+  assert (data_out == nullptr);
+  data_out_size = 0;
+
+  if (prm_get_bool_value (PRM_ID_REPLICATION_COMPRESS_PAGES_TRANSFER))
+    {
+      // it is arbitrary which of the thread entry compress structures is used in this context
+      // the thread on which this routine is to be executed is part of a thread
+      // pool which is only used to dispatch responses from the page server to the transaction servers
+      if (thread_r.log_zip_undo == nullptr)
+	{
+	  thread_r.log_zip_undo = log_zip_alloc (IO_PAGESIZE);
+	}
+
+      const bool compressed = log_zip (thread_r.log_zip_undo, data_in_size, data_in);
+      assert (compressed);
```

Comment:
Please see inside of log_zip:
    compressed = false; 
...
      if (log_zip->data_length < length)
	{
	  compressed = true;
	} 
If zipped size is not less	original, it is considered failed.

Reply:
What if there is no room to compress sine it's too compact? I mean if it's try to compress and there is no exception, but the compression ratio is 1 (same as original size), this will cause the assertion failure. I think this is a legal case. I'm not sure it's the case, though.

It might be the same as razvan's feedback.

Reply:
I will adapt the "protocol" to send the uncompressed page if that is the case:
- if compression fails (resulting length same or larger), send the uncompressed page
- on the decompress side (*TS), if the received size is the same as `IO_PAGESIZE`, it means there was no compression

@wrlawodms 
Lz4 can produce outputs larger than input. The check inside `log_zip` is in this direction, if the output is larger, to let the client code know and act accordingly.

Of course there is also the very edge case where the length of the initial data (which is also saved together with the compressed data) + the length of the compressed data is still larger than the initial data's length. In which case compressing/decompressing still makes no sense.
But I think we can ignore this corner case.

Reply:
Adapted the "protocol". Updated PR description accordingly.

> Of course there is also the very edge case where the length of the initial data (which is also saved together with the compressed data) + the length of the compressed data is still larger than the initial data's length. 

This concern is already accounted for in the implementation of `log_zip` function.